core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,588 @@
+              errorCode=Errors.NONE.code))
+          }
+        } else {
+          // for followers, just return group information and let them collect their assignment
+          responseCallback(JoinGroupResult(
+            members=Map.empty,
+            memberId=memberId,
+            generationId=group.generationId,
+            subProtocol=group.subProtocol,
+            leaderId=group.leaderId,
+            errorCode=Errors.NONE.code))
+        }
+      } else {
+        val member = if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          // if the member id is unknown, register the member to the group
+          val generatedMemberId = group.generateNextMemberId
+          val member = addMember(generatedMemberId, sessionTimeoutMs, metadata, subProtocols, group)
+          maybePrepareRebalance(group)
+          member
+        } else {
+          val member = group.get(memberId)
+          if (!member.hasMetadata(metadata) || subProtocols != member.subProtocols) {
+            // existing member changed its protocol metadata
+            updateMember(group, member, metadata, subProtocols)
+            maybePrepareRebalance(group)
+            member
+          } else {
+            // existing member rejoining a group due to restabilize
+            member
+          }
+        }
+
+        member.awaitingJoinCallback = responseCallback
+
+        if (group.is(PreparingRebalance))
+          rebalancePurgatory.checkAndComplete(ConsumerGroupKey(group.groupId))
+      }
+    }
+  }
+
+  def handleSyncGroup(groupId: String,
+                      generation: Int,
+                      memberId: String,
+                      groupAssignment: Map[String, Array[Byte]],
+                      responseCallback: SyncCallback) {
+    if (!isActive.get) {
+      responseCallback(Array.empty, Errors.GROUP_COORDINATOR_NOT_AVAILABLE.code)
+    } else if (!isCoordinatorForGroup(groupId)) {
+      responseCallback(Array.empty, Errors.NOT_COORDINATOR_FOR_GROUP.code)
+    } else {
+      val group = coordinatorMetadata.getGroup(groupId)
+      if (group == null)
+        responseCallback(Array.empty, Errors.ILLEGAL_GENERATION.code)
+      else
+        doSyncGroup(group, generation, memberId, groupAssignment, responseCallback)
+    }
+  }
+
+  def doSyncGroup(group: GroupMetadata,
+                  generationId: Int,
+                  memberId: String,
+                  groupAssignment: Map[String, Array[Byte]],
+                  responseCallback: SyncCallback) {
+    group synchronized {
+      if (group.is(Dead) || !group.has(memberId)) {
+        responseCallback(Array.empty, Errors.UNKNOWN_MEMBER_ID.code)
+      } else if (generationId != group.generationId) {
+        responseCallback(Array.empty, Errors.ILLEGAL_GENERATION.code)
+      } else if (group.is(AwaitingSync)) {
+        group.get(memberId).awaitingSyncCallback = responseCallback
+        completeAndScheduleNextHeartbeatExpiration(group, group.get(memberId))
```

Comment:
Do we need to call completeAndScheduleNextHeartbeatExpiration in both doJoin and doSync? Why we need to reschedule in doSync?